vcr/holder/wallet_test.go

```diff
@@ -225,3 +199,181 @@ func TestWallet_Present(t *testing.T) {
 		})
 	})
 }
+
+func Test_wallet_Put(t *testing.T) {
+	t.Run("ok", func(t *testing.T) {
+		storageEngine := storage.NewTestStorageEngine(t)
+		store, _ := storageEngine.GetProvider("test").GetKVStore("credentials", storage.PersistentStorageClass)
+		sut := New(nil, nil, nil, nil, store)
+		expected := createCredential(vdr.TestMethodDIDA.String())
+
+		err := sut.Put(context.Background(), expected)
+		require.NoError(t, err)
+
+		list, err := sut.List(context.Background(), vdr.TestDIDA)
+		require.NoError(t, err)
+		require.Len(t, list, 1)
+		assert.Equal(t, expected.ID.String(), list[0].ID.String())
+	})
+	t.Run("duplicate credential", func(t *testing.T) {
+		storageEngine := storage.NewTestStorageEngine(t)
+		store, _ := storageEngine.GetProvider("test").GetKVStore("credentials", storage.PersistentStorageClass)
+		sut := New(nil, nil, nil, nil, store)
+		expected := createCredential(vdr.TestMethodDIDA.String())
+
+		err := sut.Put(context.Background(), expected)
+		require.NoError(t, err)
+		err = sut.Put(context.Background(), expected)
+		require.NoError(t, err)
+
+		list, err := sut.List(context.Background(), vdr.TestDIDA)
+		require.NoError(t, err)
+		require.Len(t, list, 1)
+		assert.Equal(t, expected.ID.String(), list[0].ID.String())
+		assert.Equal(t, 1, sut.Diagnostics()[0].Result(), "duplicate credential should not increment total number of credentials")
+	})
+}
+
+func Test_wallet_List(t *testing.T) {
```

Comment:
missing happy flow test. A failure in List should fail here and not in Put

Reply:
From a purist point of view, yes, but it unnecessarily complicates the test since it's going to mess with internal details of the type (querying the database, exactly in the same way `List()` will operate):
- If the storage internals of the wallet then, more of the test will have to change. While the test is about contract (if I put in an item, can I retrieve it)
- It's not a test set in which some tests are allowed to fail (e.g. "if more than 95% of the tests succeed, it passes the quality requirements); a broken unit test is a broken system. Regardless whether it is 1, 2 or 10 tests.
- Not using List() means copying its code into the test, thus duplicating it, increasing future cost change.